### PR TITLE
Add operation lock instead of locking state Enumeration

### PR DESCRIPTION
**Pull request by turboFei**
### _Why are the changes needed?_

We meet an issue that cause all the operation stuck when closing operation.

Because now all the operations try to lock a Scala Enumeration val.

And if one of them stuck, all the others will be keep stuck.

In this pr, I add a lock for each operation.
### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by turboFei** [comment]
cc @yaooqinn @pan3793 @ulysses-you

File: kyuubi-common/src/main/scala/org/apache/kyuubi/operation/AbstractOperation.scala [hunk 1]
```diff
@@ -45,7 +46,16 @@ abstract class AbstractOperation(session: Session) extends Operation with Loggin
...
-  protected def cleanup(targetState: OperationState): Unit = state.synchronized {
+  private val lock: ReentrantLock = new ReentrantLock()
+
+  protected def withLockRequired[T](block: => T): T = {
+    try {
+      lock.lock()
+      block
+    } finally lock.unlock()
+  }
+
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/operation/AbstractOperation.scala:57]: Can we place it in Utils?

Comment (turboFei) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/operation/AbstractOperation.scala]: done

File: kyuubi-common/src/main/scala/org/apache/kyuubi/Utils.scala [hunk 2]
```diff
@@ -407,4 +408,15 @@ object Utils extends Logging {
       stringWriter.toString
     }
   }
+
+  def withLockRequired[T](lock: Lock)(block: => T): T = {
+    var locked: Boolean = false
+    try {
+      lock.lockInterruptibly()
```
Comment (turboFei) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/Utils.scala:415]: how do you think about? cc @yaooqinn

Comment (turboFei) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/Utils.scala:415]: Also cc @pan3793

Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/Utils.scala:415]: The state changing can be safely used `lock`, I do not know other places. BTW, how do we handle InteruptExcetion?

Comment (turboFei) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/Utils.scala]: Reverted

**Review by yaooqinn — APPROVED** [review]
LGTM

**Comment by pan3793** [comment]
`ResourceLock` is definitely a future-proof option

https://github.com/trinodb/trino/issues/17045

**Comment by turboFei** [comment]
thanks, merged to master

there is conflicts with branch-1.7,

**Comment by pan3793** [comment]
Cherry-picked to 1.7

**Pull request closed without merging**